src/layers/misc/channelwise_mean.cu

```diff
@@ -0,0 +1,179 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "lbann/layers/misc/channelwise_mean.hpp"
+
+namespace lbann {
+
+namespace {
+
+template <El::Int block_size>
+__global__ void mean_kernel(El::Int num_channels,
+                            El::Int channel_size,
+                            El::Int width,
+                            const DataType* __restrict__ input,
+                            El::Int input_ldim,
+                            DataType* __restrict__ output,
+                            El::Int output_ldim) {
+
+  // Indices
+  const El::Int tid = threadIdx.x;
+  const El::Int gidx = threadIdx.x + blockIdx.x * blockDim.x;
+  const El::Int bidy = blockIdx.y;
+  const El::Int bidz = blockIdx.z;
+  const El::Int nthreadsx = blockDim.x * gridDim.x;
+  const El::Int nblocksy = gridDim.y;
+  const El::Int nblocksz = gridDim.z;
+
+  // Compute local contribution for each channel
+  for (El::Int col = bidz; col < width; col += nblocksz) {
+    for (El::Int channel = bidy; channel < num_channels; channel += nblocksy) {
+
+      // Sum for each thread
+      DataType private_sum = 0;
+      for (El::Int i = gidx; i < channel_size; i += nthreadsx) {
+        private_sum += input[i + channel*channel_size + col*input_ldim];
+      }
+
+      // Shared memory reduction to get sum for each block
+      /// @todo unroll loops
+      __shared__ DataType shared_sums[block_size];
+      shared_sums[tid] = private_sum;
+      for (El::Int stride = block_size / 2; stride > 0; stride /= 2) {
```

Comment:
This traversal pattern could do with a quick comment, it's not especially clear what's going on.

Reply:
Cub has some primitives that hide dirty details of reductions.